### PR TITLE
C++ comparison of upscaling .txt files 

### PR DESCRIPTION
**Pull request by kristfho**
This test run the upscaling binaries and compares the resulting txt file with the reference solution located at `~/opm-upsacaling/tests/input_data/reference_solution/`. The test compares the files value by value and throws a exception when the deviation between the values exceed a specified limit.

### CODE REVIEW

File: tests/compareUpscaling.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <sstream>
+#include <string>
+#include <vector>
+#include <cstdlib>
+#include <cmath>
+#include <opm/common/ErrorMacros.hpp>
+
+
+double getAbsDeviation(double value1, double value2){
+  return abs(value1-value2);
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: std::abs

File: compareUpscaling.cmake [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                         27cellsAniso stoneAniso.txt 30 8)
+
+if((DUNE_ISTL_VERSION_MAJOR GREATER 2) OR
+   (DUNE_ISTL_VERSION_MAJOR EQUAL 2 AND DUNE_ISTL_VERSION_MINOR GREATER 2))
+  add_dependencies (test-suite upscale_elasticity)
+  add_test_upscale_elasticity(EightCells mpc)
+  add_test_upscale_elasticity(EightCells mortar)
+endif()
+
+add_custom_target(clear_test_data}_asciiComparison ${CMAKE_COMMAND} -E remove_directory ${RESULT_PATH}
```
Comment (akva2) [inline compareUpscaling.cmake]: move to shell script

File: tests/compareUpscaling.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+void readData(const char* fileName, std::vector<double> &dataVec){
+  std::ifstream inFile (fileName, std::ifstream::in);
+  std::string line, value; 
+
+  while(std::getline(inFile, line)){
+    std::stringstream ss(line);
+    while(ss >> value){
+      if(value == "#" || value.substr(0,1) == "#" ){break;}
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: substr suffices.

File: tests/compareUpscaling.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  }
+  
+  double absDev, relDev; 
+  for (int i = 0; i < testVec.size(); i++){
+    absDev = getAbsDeviation(testVec[i], compareVec[i]);
+    relDev = getRelDeviation(testVec[i], compareVec[i]);
+    if (absDev > absoluteTolerance){
+      std::cout << "For data value " << i << ". The absolute deviation is "<< absDev << " greater than the limit "<< absoluteTolerance << std::endl; 
+      OPM_THROW(std::invalid_argument, "The absolute deviation exceeded the limit.");
+	}
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: indent

File: tests/compareUpscaling.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    
+    if (relDev > relativeTolerance && absDev > 1e-6){
+      std::cout << "For data value " << i << "- The relative deviation is "<< relDev << " greater than the limit "<< relativeTolerance << std::endl; 
+      OPM_THROW(std::invalid_argument, "The relative deviation exceeded the limit.");
+	}
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: indent

File: tests/run-compareUpscaling.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+set -e
+
+INPUT_DATA_PATH="$1"
+RESULT_PATH="$2"
+BINPATH="$3"
+TEST_NAME="$4"
+ABS_TOL="$5"
+REL_TOL="$6"
+EXE_NAME="$7"
+TEST_ARGS="${@:8:100}"
```
Comment (akva2) [inline tests/run-compareUpscaling.sh]: shift

File: tests/compareUpscaling.cpp [hunk 7]
```diff
@@ -8,78 +26,78 @@
...
     }
-  }
 
 }
 
 void startTest(const char* testFile, const char* compareFile,  double absoluteTolerance, double relativeTolerance){
-  std::vector<double> testVec, compareVec;
+    std::vector<double> testVec, compareVec;
+    std::cout << " MIN TEST! " << std::endl;
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: no.

File: tests/compareUpscaling.cpp [hunk 8]
```diff
@@ -8,78 +26,78 @@
...
+	if (absDev > absoluteTolerance){
+	    std::cout << "For data value " << i << ". The absolute deviation is "<< absDev << " greater than the limit "<< absoluteTolerance << std::endl; 
+	    OPM_THROW(std::invalid_argument, "The absolute deviation exceeded the limit.");
 	}
     
-    if (relDev > relativeTolerance && absDev > 1e-6){
-      std::cout << "For data value " << i << "- The relative deviation is "<< relDev << " greater than the limit "<< relativeTolerance << std::endl; 
-      OPM_THROW(std::invalid_argument, "The relative deviation exceeded the limit.");
+	if (relDev > relativeTolerance && absDev > 1e-6){
+	    for(size_t i = 0; i < testVec.size(); i++){
```
Comment (akva2) [inline tests/compareUpscaling.cpp]: leftovers.

Comment (kristfho) [inline tests/compareUpscaling.cpp]: leftovers?

File: CMakeLists_files.cmake [hunk 9]
```diff
@@ -47,7 +47,7 @@ list(APPEND MAIN_SOURCE_FILES
 	opm/porsol/euler/ImplicitCapillarity.cpp
 	opm/upscaling/ParserAdditions.cpp
 	opm/upscaling/RelPermUtils.cpp
-        opm/upscaling/initCPGrid.cpp
+  opm/upscaling/initCPGrid.cpp
```
Comment (akva2) [inline CMakeLists_files.cmake]: whoops.

**Comment by akva2** [comment]
jenkins build this please

**Pull request merged**